src/discrete_alloc.rs

```diff
@@ -0,0 +1,271 @@
+use std::alloc::{self, Layout};
+use std::sync;
+
+use nix::sys::mman::ProtFlags;
+
+use crate::helpers::ToU64;
+
+static ALLOCATOR: sync::Mutex<MachineAlloc> = sync::Mutex::new(MachineAlloc::empty());
+
+/// A distinct allocator for the `MiriMachine`, allowing us to manage its
+/// memory separately from that of Miri itself.
+#[derive(Debug)]
+pub struct MachineAlloc {
+    pages: Vec<*mut u8>,
+    huge_allocs: Vec<(*mut u8, usize)>,
+    allocated: Vec<Box<[u8]>>,
+    page_size: usize,
+    enabled: bool,
+}
+
+// SAFETY: We only point to heap-allocated data
+unsafe impl Send for MachineAlloc {}
+
+impl MachineAlloc {
+    // Allocation-related methods
+
+    /// Initializes the allocator with placeholder 4k pages.
```

Comment:
I don't understand the comment. This doesn't seem to actually put any placeholder pages into the allocator?